JIT: Fix build on GCC

trungnt2910 opened the pull request:
Otherwise, building `compiler.cpp` would trigger a zero-length format string error.

https://github.com/dotnet/runtime/blob/4e86b1c63d9c41c6bfb6f42710be907199ce2671/src/coreclr/jit/compiler.cpp#L5253

am11 commented:
Which version of gcc it is failing with? We have gcc 13 leg, which was passing.

Can we not suppress it and instead fix the issue?

trungnt2910 commented:
Tried building with
```
 ./build.sh --ci -c Release -arch x64 -os linux -gcc /p:OfficialBuildId=20240307.1
```
on GCC:
```
gcc (Ubuntu 12.3.0-1ubuntu1~23.04) 12.3.0
```

Pretty sure the error occurs on [GCC 13](https://godbolt.org/z/W5naKjMjo) as well.
It also affects the [Haiku cross-build of .NET on GCC 13.2.0](https://github.com/trungnt2910/dotnet-builds/actions/runs/8182363798/job/22373597739).

am11 commented:
This is gcc 13.2 build from main branch https://dev.azure.com/dnceng-public/public/_build/results?buildId=592829&view=logs&jobId=9d44ffa5-6dbb-5c44-43bb-984541de6129&j=9d44ffa5-6dbb-5c44-43bb-984541de6129&t=2c964513-cc68-49c8-ba29-8574bb6d43f1 but we use -checked build (instead of -release) which apparently is not building that code.

The fix is:
```diff
--- a/src/coreclr/jit/compiler.cpp
+++ b/src/coreclr/jit/compiler.cpp
@@ -5241,7 +5241,7 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
 #elif FEATURE_SIMD
         fprintf(compJitFuncInfoFile, " %s\n", eeGetMethodFullName(info.compMethodHnd));
 #endif
-        fprintf(compJitFuncInfoFile, ""); // in our logic this causes a flush
+        fprintf(compJitFuncInfoFile, "%s", ""); // in our logic this causes a flush
     }
 #endif // FUNC_INFO_LOGGING
 }
```

jakobbotsch commented:
> The fix is:
> 
> ```diff
> --- a/src/coreclr/jit/compiler.cpp
> +++ b/src/coreclr/jit/compiler.cpp
> @@ -5241,7 +5241,7 @@ void Compiler::compCompile(void** methodCodePtr, uint32_t* methodCodeSize, JitFl
>  #elif FEATURE_SIMD
>          fprintf(compJitFuncInfoFile, " %s\n", eeGetMethodFullName(info.compMethodHnd));
>  #endif
> -        fprintf(compJitFuncInfoFile, ""); // in our logic this causes a flush
> +        fprintf(compJitFuncInfoFile, "%s", ""); // in our logic this causes a flush
>      }
>  #endif // FUNC_INFO_LOGGING
>  }
> ```

That would break the purpose of the call:

https://github.com/dotnet/runtime/blob/5373c0113e55727ed8e1b9af5f44f7779bf5089c/src/coreclr/jit/error.cpp#L314-L321

jakobbotsch commented:
The call can probably just be replaced by `fflush(compJitFuncInfoFile)`. In fact I see other places in the source already doing exactly that.

trungnt2910 committed:
JIT: Fix build on GCC

Flush the file with `fflush` instead of a `fprintf` call with an empty
string. This fixes the GCC zero-length format string warning.
